### ISSUE TITLE
All Modules :: Content Of Heading FIeld Is Lost When Content Contains HTML Entities

### ISSUE THREAD

**Issue by lots0logs** [elegantthemes/Divi-Beta#829]
### Steps To Reproduce:
1.  Add the following into a heading field: `antergos ❤ numix`
2. Save the module settings.
3. Reopen the module settings.
4. Heading is gone. 

## 📎 Attached PR
- https://github.com/elegantthemes/submodule-builder/pull/1507

**Comment by fikrirasyid** [comment]
@lots0logs can you elaborate how to replicate this issue? Screencast would be really helpful as well. I tried replicating the issue but it appears to be working as expected. This is how the visual builder page looks like after being refreshed:

![screen shot 2016-12-08 at 12 13 25 pm](https://cloud.githubusercontent.com/assets/916442/20998353/c2df0e10-bd3f-11e6-99e0-d260072de24b.png)

I cross-checked it on the frontend page and it appears to be properly saved and displayed as well:

![screen shot 2016-12-08 at 12 13 32 pm](https://cloud.githubusercontent.com/assets/916442/20998360/d09b19fe-bd3f-11e6-8d6b-850b11731b7a.png)

**Comment by lots0logs** [comment]
@fikrirasyid This happens in the BB. Sorry I should have been more specific! :doh:

**Comment by fikrirasyid** [comment]
@lots0logs i tried doing it via BB and it still works as expected tho:

![screen shot 2016-12-09 at 8 43 48 am](https://cloud.githubusercontent.com/assets/916442/21034604/a9fa1e1a-bdeb-11e6-915b-5d8c11d60cdf.png)
*Create new page and adding the emoji contained text via copy paste and OSX emoji picker and it still works as expected*

Can you record a screencast? I think I missed some steps here 🤔

**Comment by lots0logs** [comment]
@fikrirasyid Hmm..it might be because I typed the HTML entity code directly in the module. Github converted it to the symbol when I copy/pasted it (and I didnt notice until now). This is what I put into the module: `antergos &#10084; numix`

**Comment by fikrirasyid** [comment]
@lots0logs I've tried typing it manually and it still works on most module's header. It mostly fails on module item and description's text tho. Is this what you experienced too?

![screen_shot_2016-12-09_at_3_03_49_pm](https://cloud.githubusercontent.com/assets/916442/21042345/65e1ee6e-be23-11e6-84f1-7ded3beb97bd.jpg)

![screen_shot_2016-12-09_at_3_03_45_pm](https://cloud.githubusercontent.com/assets/916442/21042346/65e3d436-be23-11e6-876f-5a52cf28234a.jpg)

**Comment by lots0logs** [comment]
@fikrirasyid Yeah its a slide..you can see it [here](https://antergos.com) (the 2nd slide).

**Comment by fikrirasyid** [comment]
@lots0logs alright, thank you for confirming it. Can you clarify one more thing? So I've found out that:

1. Create new slider module
2. Add slide item. Type the `antergos &#10084; numix` on header field
3. Save module. Save page
4. You'll see that the heart icon is actually saved properly
5. Go back to editor. Open the slider module. Now you'll see that the previous header is gone
6. If you still see the heart icon, try saving/closing the slider module modal, then open the slider modal again

**Comment by lots0logs** [comment]
@fikrirasyid Yes that's correct :+1:

**Comment by timayhelen** [comment]
@fikrirasyid Once an emoji has been saved in the BB, and then loaded in the VB, if you edit the header that contains the emoji and try to save, the emoji will be removed.

![emoji](https://cloud.githubusercontent.com/assets/3333028/21124019/29a211ac-c091-11e6-95e4-c8414e89448c.gif)

**Comment by lots0logs** [comment]
![homer-simpson-doh](https://cloud.githubusercontent.com/assets/4675662/21124158/ca8c7fce-c0a2-11e6-8a5d-2abd5d899073.gif)

**Comment by fikrirasyid** [comment]
@timayhelen  @lots0logs I've pushed https://github.com/elegantthemes/submodule-builder/commit/21ee24b540a03939cda0ab913bcde8f9ab6453da which should fix the issue. The solution used on visual builder is basically the similar fix added on BB because essentially, the issue remains the same: WordPress automatically staticize emoji / html entities with `<img />` markup. This causes react escapes the staticized emoji. This can be "fixed" by disabling WordPress' emoji detection script just like what we does on BB. There are one caveat tho. Since builder doesn't replaced the emoji with staticized emoji, it might look little inconsistent since the emoji uses its actual output. For example, here's how inserting emoji using emoji and HTML entities looks like on VB:

![screen shot 2016-12-13 at 3 30 29 pm](https://cloud.githubusercontent.com/assets/916442/21134447/81487b70-c150-11e6-92cd-061393fc9ba4.png)

Here's how it looks on VB after the page is saved and reloaded:

![screen shot 2016-12-13 at 3 30 54 pm](https://cloud.githubusercontent.com/assets/916442/21134587/2a82fcf6-c151-11e6-8878-9c5199455a38.png)

However, the icon looks consistent on actual front-end page, since they are automatically replaced by staticized emoji:

<img width="825" alt="screen shot 2016-12-13 at 4 29 20 pm" src="https://cloud.githubusercontent.com/assets/916442/21134754/d9539a92-c151-11e6-94cf-8d61ed0baf32.png">

**Comment by lots0logs** [comment]
@fikrirasyid Is there no reasonable way we can get the VB to render the HTML entity code before the page is saved and refreshed?

**Comment by fikrirasyid** [comment]
@lots0logs do you mean to avoid this left scenario?

![screen shot 2016-12-13 at 3 30 29 pm](https://cloud.githubusercontent.com/assets/916442/21134447/81487b70-c150-11e6-92cd-061393fc9ba4.png)

**Comment by lots0logs** [comment]
@fikrirasyid Yes, exactly. But keep in mind I said ***reasonable***, (its okay if the answer is no).

**Comment by fikrirasyid** [comment]
@lots0logs I couldn't find a proper solution for it the first time, but I'll take a second look on it. I agree that displaying html entities doesn't look and feel really good 🤔

**Comment by fikrirasyid** [comment]
@lots0logs I finally found a way to automatically convert HTML entities into its respective character:

![829-pr](https://cloud.githubusercontent.com/assets/916442/21269675/4143124a-c3e6-11e6-8cb7-7473345d20db.gif)

Code-wise, it turns out to be amazingly straightforward. Number-based HTML entities are relatively simple to be handled using `String.fromCharCode()` (see: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/fromCharCode). So what we're doing is basically:

```
var value = event.target.value.replace(/&#(\d+);/g, function(match, dec) {
  return String.fromCharCode(dec);
});
```

There are two buts tho:

1. It makes me nervous about XSS attack even though I realize that it should be safe since the output returned data of the builder is shortcode that are being passed into `wp_update_post()` that is sanitized via `sanitize_post()`. Which means theoretically, bad script wouldn't reach database however do you see any possibility that this is being exploited? I've tried doing simple XSS myself but everything looks good tho.
2. I learn this approach via this article: http://ourcodeworld.com/articles/read/188/encode-and-decode-html-entities-using-pure-javascript. If my understanding toward our 3rd party code policy is correct, we might need to attribute the author for his snippet. Unfortunately, there is no explicit license mentioned on his site.  What do you think, @timayhelen I've contacted the author regarding this issue and still waiting for response.

Considering point no. 2, I haven't commit and push the changes. Let me know what you guys think. Is it worth pursuing or let's merge with what we have right now.

**Comment by lots0logs** [comment]
@fikrirasyid 

1. I don't think there is any risk of XSS in this case. @Jronk what do you think?
2. The author of that post you linked included this line below the code so I think we are okay:
   >Note : feel free to copy every single function and include it in your project as you wish.

**Comment by Jronk** [comment]
```
jQuery('body').append('<script>' + String.fromCharCode(97, 108, 101, 114, 116, 40, 34, 104, 105, 34, 41, 59) + '</script>');

htmlentities.decode('&#97;&#108;&#101;&#114;&#116;&#40;&#34;&#104;&#105;&#34;&#41;&#59;');
```

**Comment by fikrirasyid** [comment]
@lots0logs @Jronk I've pushed a commit which fires automatic html entities conversion on input field. The conversion is triggered when cursor leaves the input field or `enter` is pressed while focusing on input text. Here's how it looks like:

![829-commit](https://cloud.githubusercontent.com/assets/916442/21284732/6753ddea-c457-11e6-9135-89b51c987dcc.gif)

> The author of that post you linked included this line below the code so I think we are okay:

I just saw it now :facepalm: You're right about that. The author has responded to my question in similar fashion:

![screen shot 2016-12-17 at 12 51 36 pm](https://cloud.githubusercontent.com/assets/916442/21284736/98e1a22a-c457-11e6-9ab0-18905f1b87a8.png)
